### PR TITLE
Refactor admission controller contructor with Config.Complete.New flow

### PR DESCRIPTION
**Pull request by yue9944882**
**What this PR does / why we need it**:

The parameter list for `BuildAdmissionPluginInitializers ` is too long and meaningless. Gathering information for creating admission controller into `AdmissionConfig` just like we do for `AuthenticationConfig`/`AuthorizationConfig`.

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/kube-apiserver/app/server.go [hunk 1]
```diff
@@ -536,36 +512,15 @@ func buildGenericConfig(
...
-			},
-		}
+	admissionConfig := &kubeapiserveradmission.AdmissionConfig{
+		EnableAggregatorRouting: s.EnableAggregatorRouting,
+		CloudConfigFile:         s.CloudProvider.CloudConfigFile,
+		ExternalInformers:       versionedInformers,
+		InternalInformers:       sharedInformers,
+		LoopbackClientConfig:    genericConfig.LoopbackClientConfig,
 	}
-	pluginInitializers, admissionPostStartHook, err = BuildAdmissionPluginInitializers(
```
Comment (yue9944882) [inline cmd/kube-apiserver/app/server.go]: before

File: cmd/kube-apiserver/app/server.go [hunk 2]
```diff
@@ -536,36 +512,15 @@ func buildGenericConfig(
...
-		s,
-		client,
-		sharedInformers,
-		serviceResolver,
-		webhookAuthResolverWrapper,
-	)
+
+	pluginInitializers, admissionPostStartHook, err = admissionConfig.Complete(proxyTransport).New()
```
Comment (yue9944882) [inline cmd/kube-apiserver/app/server.go]: after😄

File: pkg/kubeapiserver/admission/config.go [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+	LoopbackClientConfig *rest.Config
+
+	ServiceResolver            aggregatorapiserver.ServiceResolver
+	WebhookAuthResolverWrapper config.AuthenticationInfoResolverWrapper
+
+	ExternalInformers externalinformers.SharedInformerFactory
+	InternalInformers internalinformers.SharedInformerFactory
+}
+
+func (c *AdmissionConfig) Complete(proxyTransport *http.Transport, enabledAggregatorRouting bool) completedAdmissionConfig {
```
Comment (sttts) [inline pkg/kubeapiserver/admission/config.go]: `Complete` usually usually only sets fields which are nil. Here though fields are unconditionally overridden. This looks more like a constructor than completion.

File: pkg/kubeapiserver/admission/config.go [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+package admission
+
+import (
+	"net/http"
+	"net/url"
+	"time"
+
+	"github.com/golang/glog"
+	"io/ioutil"
```
Comment (sttts) [inline pkg/kubeapiserver/admission/config.go]: nit: import order

File: pkg/kubeapiserver/admission/config.go (hunk 3 above)
Comment (yue9944882) [reply pkg/kubeapiserver/admission/config.go]: yeah.. added checking nil before setting fields

**Comment by yue9944882** [comment]
@sttts frankly, think it twice and am not even convinced by this change myself. 😅obviously this pull makes the code much cleaner than before, but it doesn't really help clarifying the dependency of instantiating `pluginInitializer` and `admissionPostStartHook`. as a matter of fact, we're still collecting information here and there and simply lumping an `admissionConfig`.

**Comment by sttts** [comment]
> am not even convinced by this change myself. 😅obviously this pull makes the code much cleaner than before, but it doesn't really help clarifying the dependency of instantiating pluginInitializer and admissionPostStartHook. as a matter of fact, we're still collecting information here and there and simply lumping them into an admissionConfig.
> 
> Merge state

Complex plumbing will stay complex, no matter how you move parts of it around :)

**Comment by deads2k** [comment]
> @sttts frankly, think it twice and am not even convinced by this change myself. obviously this pull makes the code much cleaner than before, but it doesn't really help clarifying the dependency of instantiating pluginInitializer and admissionPostStartHook. as a matter of fact, we're still collecting information here and there and simply lumping them into an admissionConfig.

Make the best lemonade you can.  Forcing structure here will help you tease apart dependency ordering and it's ok to do it step-wise as long as you're actually coming back to it.

File: pkg/kubeapiserver/admission/config.go (hunk 3 above)
Comment (yue9944882) [reply pkg/kubeapiserver/admission/config.go]: umm.. different from the fields inside `AdmissionConfig`, `proxyTransport` and `enabledAggregatorRouting` are sth/information extracted from aggregation layer. it's not an inside part of admission logic but more like an outside wrapper for proxy usage. an `AdmissionConfig` shouldn't contain extra information about aggregator, amiright?

Comment (yue9944882) [reply pkg/kubeapiserver/admission/config.go]: or what about move it as `New(proxyTransport, enabledAggregatorRouting)`?

Comment (sttts) [reply pkg/kubeapiserver/admission/config.go]: `New(proxyTransport, enabledAggregatorRouting)` is what I meant, yes.

File: pkg/kubeapiserver/admission/config.go [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+	WebhookAuthResolverWrapper config.AuthenticationInfoResolverWrapper
+
+	ExternalInformers externalinformers.SharedInformerFactory
+	InternalInformers internalinformers.SharedInformerFactory
+}
+
+func (c *AdmissionConfig) withAggregatorProxy(proxyTransport *http.Transport, enabledAggregatorRouting bool) {
+
+	if enabledAggregatorRouting {
+		c.ServiceResolver = aggregatorapiserver.NewEndpointServiceResolver(
```
Comment (sttts) [inline pkg/kubeapiserver/admission/config.go]: do we have to store ServiceResolver at all in the config? Isn't it transient and not used anywhere from the config?

File: pkg/kubeapiserver/admission/config.go [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+		c.ServiceResolver = aggregatorapiserver.NewClusterIPServiceResolver(
+			c.ExternalInformers.Core().V1().Services().Lister(),
+		)
+	}
+	// resolve kubernetes.default.svc locally
+	if localHost, err := url.Parse(c.LoopbackClientConfig.Host); err == nil {
+		c.ServiceResolver = aggregatorapiserver.NewLoopbackServiceResolver(c.ServiceResolver, localHost)
+	}
+
+	c.WebhookAuthResolverWrapper = func(delegate webhookconfig.AuthenticationInfoResolver) webhookconfig.AuthenticationInfoResolver {
```
Comment (sttts) [inline pkg/kubeapiserver/admission/config.go]: same for this

**Comment by sttts** [comment]
lgtm (when tests are green)
/approve

**Comment by yue9944882** [comment]
@sttts the tests are never green 😭must be sth wrong in this pull. 
am seriously taking a look..

**Comment by caesarxuchao** [comment]
I think it's related to the serviceResolver plumbing. See the apiserver [log](https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/67133/pull-kubernetes-e2e-gce/49363/artifacts/e2e-49363-674b9-master/kube-apiserver.log),
```
I0822 07:26:12.873122       1 wrap.go:42] GET /apis?timeout=32s: (587.995µs) 200 [kube-controller-manager/v1.13.0 (linux/amd64) kubernetes/446af58/system:serviceaccount:kube-system:generic-garbage-collector [::1]:53098]
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x1573172]

goroutine 30868 [running]:
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAudit.func1.1(0xc42e350500, 0x7f0e905dbd20, 0xc420ee99e0, 0xc420304eb0, 0x1, 0x1, 0x0, 0x0)
	staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:84 +0x1f8
panic(0x35dafc0, 0x718c000)
	GOROOT/src/runtime/panic.go:502 +0x229
k8s.io/kubernetes/vendor/k8s.io/kube-aggregator/pkg/apiserver.(*proxyHandler).ServeHTTP(0xc423945c70, 0x52df0a0, 0xc431fc19e8, 0xc4345c8900)
	staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:113 +0x282
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*pathHandler).ServeHTTP(0xc427b9cc00, 0x52df0a0, 0xc431fc19e8, 0xc4345c8900)
	staging/src/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:241 +0x514
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/mux.(*PathRecorderMux).ServeHTTP(0xc421b24930, 0x52df0a0, 0xc431fc19e8, 0xc4345c8900)
	staging/src/k8s.io/apiserver/pkg/server/mux/pathrecorder.go:234 +0x8a
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server.director.ServeHTTP(0x3dcdded, 0xf, 0xc42aa2c990, 0xc421b24930, 0x52df0a0, 0xc431fc19e8, 0xc4345c8900)
	staging/src/k8s.io/apiserver/pkg/server/handler.go:154 +0x6a8
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthorization.func1(0x52df0a0, 0xc431fc19e8, 0xc4345c8900)
	staging/src/k8s.io/apiserver/pkg/endpoints/filters/authorization.go:64 +0x4b1
net/http.HandlerFunc.ServeHTTP(0xc42887e780, 0x52df0a0, 0xc431fc19e8, 0xc4345c8900)
	GOROOT/src/net/http/server.go:1947 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.WithMaxInFlightLimit.func1(0x52df0a0, 0xc431fc19e8, 0xc4345c8900)
	staging/src/k8s.io/apiserver/pkg/server/filters/maxinflight.go:160 +0x3fc
net/http.HandlerFunc.ServeHTTP(0xc427874420, 0x52df0a0, 0xc431fc19e8, 0xc4345c8900)
	GOROOT/src/net/http/server.go:1947 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithImpersonation.func1(0x52df0a0, 0xc431fc19e8, 0xc4345c8900)
	staging/src/k8s.io/apiserver/pkg/endpoints/filters/impersonation.go:50 +0x1ce1
net/http.HandlerFunc.ServeHTTP(0xc42887e800, 0x52df0a0, 0xc431fc19e8, 0xc4345c8900)
	GOROOT/src/net/http/server.go:1947 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAudit.func1(0x7f0e8fd1e178, 0xc431fc19e0, 0xc4345c8800)
	staging/src/k8s.io/apiserver/pkg/endpoints/filters/audit.go:106 +0x4b7
net/http.HandlerFunc.ServeHTTP(0xc42887e840, 0x7f0e8fd1e178, 0xc431fc19e0, 0xc4345c8800)
	GOROOT/src/net/http/server.go:1947 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.WithAuthentication.func1(0x7f0e8fd1e178, 0xc431fc19e0, 0xc4345c8800)
	staging/src/k8s.io/apiserver/pkg/endpoints/filters/authentication.go:75 +0x2b3
net/http.HandlerFunc.ServeHTTP(0xc42887e8c0, 0x7f0e8fd1e178, 0xc431fc19e0, 0xc4345c8700)
	GOROOT/src/net/http/server.go:1947 +0x44
k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP.func1(0xc4224ca040, 0x52ea3e0, 0xc431fc19e0, 0xc4345c8700, 0xc421b76d20)
	staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go:97 +0x8d
created by k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/server/filters.(*timeoutHandler).ServeHTTP
	staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go:96 +0x1b5
```

In which `staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go:113` is dereferencing the serviceResolver.

The cause is that after your refactor, ` func buildGenericConfig` doesn't initialize the `serviceResolver` anymore. However, the resolver is later used to construct the aggregator, thus the nil pointer reference.

**Comment by yue9944882** [comment]
@caesarxuchao thanks for the hint 👍🏻i'm pretty it's now working now 🙂

**Comment by yue9944882** [comment]
/cc @caesarxuchao 
plitely bumping this thread

File: pkg/kubeapiserver/admission/config.go [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+		if err != nil {
+			glog.Fatalf("Error reading from cloud configuration file %s: %#v", c.CloudConfigFile, err)
+		}
+	}
+	internalClient, err := internalclientset.NewForConfig(c.LoopbackClientConfig)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	webhookPluginInitializer := webhookinit.NewPluginInitializer(webhookAuthResolverWrapper, serviceResolver)
```
Comment (caesarxuchao) [inline pkg/kubeapiserver/admission/config.go]: nit: move this line to where webhookAuthResolverWrapper is defined

File: pkg/kubeapiserver/admission/config.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+		quotainstall.NewQuotaConfigurationForAdmission(),
+	)
+
+	admissionPostStartHook := func(context genericapiserver.PostStartHookContext) error {
+		discoveryRESTMapper.Reset()
+		go utilwait.Until(discoveryRESTMapper.Reset, 30*time.Second, context.StopCh)
+		return nil
+	}
+
+	return []admission.PluginInitializer{kubePluginInitializer, webhookPluginInitializer}, admissionPostStartHook, nil
```
Comment (caesarxuchao) [inline pkg/kubeapiserver/admission/config.go]: Why do you switch the order between kubePluginInitializer and webhookPluginInitializer? I don't know if the order is significant, but keeping the original order makes review easier :)

**Review by caesarxuchao — COMMENTED** [review]
Two nits and otherwise lgtm. Please ping me on slack when it's ready.

File: pkg/kubeapiserver/admission/config.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	internalinformers "k8s.io/kubernetes/pkg/client/informers/informers_generated/internalversion"
+	quotainstall "k8s.io/kubernetes/pkg/quota/v1/install"
+)
+
+type AdmissionConfig struct {
+	CloudConfigFile string
+
```
Comment (caesarxuchao) [inline pkg/kubeapiserver/admission/config.go]: Remove the extra whiteline.

File: pkg/kubeapiserver/admission/config.go [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+				}
+				return ret, err
+			},
+		}
+	}
+	return webhookAuthResolverWrapper
+}
+
+func (c *AdmissionConfig) New(proxyTransport *http.Transport, serviceResolver webhook.ServiceResolver) ([]admission.PluginInitializer, server.PostStartHookFunc, error) {
+
```
Comment (caesarxuchao) [inline pkg/kubeapiserver/admission/config.go]: Remove the extra whiteline.

File: pkg/kubeapiserver/admission/config.go [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+	"net/http"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/api/core/v1"
+	utilwait "k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apiserver/pkg/admission"
+	webhookinit "k8s.io/apiserver/pkg/admission/plugin/webhook/initializer"
+
```
Comment (caesarxuchao) [inline pkg/kubeapiserver/admission/config.go]: Remove the white line.

File: pkg/kubeapiserver/admission/config.go [hunk 12]
```diff
@@ -0,0 +1,112 @@
...
+	"k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	internalinformers "k8s.io/kubernetes/pkg/client/informers/informers_generated/internalversion"
+	quotainstall "k8s.io/kubernetes/pkg/quota/v1/install"
+)
+
+type AdmissionConfig struct {
+	CloudConfigFile string
+
+	LoopbackClientConfig *rest.Config
+
```
Comment (caesarxuchao) [inline pkg/kubeapiserver/admission/config.go]: Here as well :)

**Review by caesarxuchao — COMMENTED** [review]
lgtm. Small nits.

**Pull request merged**